### DISCUSSION TITLE
Const Tuples!

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#2236] [Language Ideas]
C# gives an error for this:
`const (string, int) x = ("a", 1);`
saying:
> The type '(string, int)' cannot be declared const

I suggest to allow us to define const tuples. If CLR cann't accept this, C# can compile the previous code to:
```
const string x_item1 = "a";
const int x_item2 = 1;
```

Same for:
`const (string name, int age) x = ("Adam", 20);`

Which can be converted to:
```
const string x_name = "Adam";
const int x_age = 20;
```

**Comment by HaloFour** [comment]
> If CLR cann't accept this

The CLR can't do this.

> C# can compile the previous code to:

What would be the point of that?  How and where would you use this "tuple" that you need special language support for it, especially when it wouldn't be a tuple?

**Reply by Aniobodo** [reply]
I would use it to hold related values. 
For instance, to hold limit values: (int Min, int Max) limit = (10, 20);
It looks more elegant with tuple, instead of: limitMin = 10; limitMax = 20;

**Reply by HaloFour** [reply]
```cs
static class Limit {
    const int Min = 10;
    const int Max = 20;
}
```

**Comment by ghost** [comment]
@HaloFour
I will use this const tuple as a tuple, but C# will convert all that to individual consts :)
Or better: make CLR accept this!
Tuples are powerful, and we want to use them in all situations. In fact this proposal came from this issue:
https://github.com/dotnet/vblang/issues/372

**Comment by ghost** [comment]
And this is also related to this proposal: https://github.com/dotnet/csharplang/issues/2237

**Comment by HaloFour** [comment]
> Or better: make CLR accept this!

Even if this was considered it would never be supported across all CLRs.  The .NET Framework is pretty much guaranteed to not be getting new features at this point.

> Tuples are powerful, and we want to use them in all situations.

That's not a reason.  You're free to use them as not constants, such as with `public readonly` fields.

**Comment by ghost** [comment]
@HaloFour  
ReadOnly members can't be defined in functions. Consts can.

**Comment by HaloFour** [comment]
https://github.com/dotnet/csharplang/issues/188

And that requires no CLR changes.

**Comment by ghost** [comment]
@HaloFour 
Nice. But if that happened, what would be the difference between const and readonly keywords? Hence, what prevents the compiler then from compiling const expr to readonly expr under the hood? Just asking.

**Comment by HaloFour** [comment]
Consts are more like aliases.  When you use the const anywhere in your code the compiler replaces the name of the const with the value.  Const locals are really a compiler trick.  Const fields are *mostly* a CLR feature, literal fields.  The names exist in metadata but the fields cannot be referenced directly.  If you reference one of these members the compiler must interpret the binary value of the field out of the assembly metadata and embed that value directly in your code.

I say "mostly" because the CLR doesn't support literal fields of `decimal`.  The C# compiler fakes it by emitting a real readonly `decimal` field and marking it with an attribute `DecimalConstant` which contains the binary data of that decimal.

Technically the compiler could do that for any blittable type, and it's been brought up before.  The question is whether its worth it.  You don't really gain anything over a `readonly` field.
